contrib/win32/win32compat/ssh-agent/keyagent-request.c

```diff
@@ -367,6 +541,173 @@ process_remove_all(struct sshbuf* request, struct sshbuf* response, struct agent
 	return r;
 }
 
+#ifdef ENABLE_PKCS11
+int process_add_smartcard_key(struct sshbuf* request, struct sshbuf* response, struct agent_connection* con)
+{
+	char *provider = NULL, *pin = NULL, canonical_provider[PATH_MAX];
+	int i, count = 0, r = 0, request_invalid = 0, success = 0;
+	struct sshkey **keys = NULL;
+	struct sshkey* key = NULL;
+	size_t pubkey_blob_len, provider_len, pin_len, epin_len;
+	u_char *pubkey_blob = NULL;
+	char *thumbprint = NULL;
+	char *epin = NULL;
+	HKEY reg = 0, sub = 0, user_root = 0;
+	SECURITY_ATTRIBUTES sa = { 0, NULL, 0 };
+
+	if ((r = sshbuf_get_cstring(request, &provider, &provider_len)) != 0 ||
+		(r = sshbuf_get_cstring(request, &pin, &pin_len)) != 0) {
+		debug("add smartcard request is invalid");
+		request_invalid = 1;
+		goto done;
+	}
+
+	if (realpath(provider, canonical_provider) == NULL) {
+		debug("failed PKCS#11 add of \"%.100s\": realpath: %s",
+			provider, strerror(errno));
+		request_invalid = 1;
+		goto done;
+	}
+
+	// Remove 'drive root' if exists
+	if (canonical_provider[0] == '/')
+		memmove(canonical_provider, canonical_provider + 1, strlen(canonical_provider));
+	if (get_user_root(con, &user_root) != 0 ||
+		is_reg_sub_key_exists(user_root, SSH_PKCS11_PROVIDERS_ROOT, canonical_provider))
+		goto done;
+
+	pkcs11_init(0);
+
+	count = pkcs11_add_provider(canonical_provider, pin, &keys, NULL);
+	if (count <= 0) {
+		debug("failed to add key to store");
+		goto done;
+	}
+	for (i = 0; i < count; i++) {
+		key = keys[i];
+		memset(&sa, 0, sizeof(SECURITY_ATTRIBUTES));
+		sa.nLength = sizeof(sa);
+		if ((!ConvertStringSecurityDescriptorToSecurityDescriptorW(REG_KEY_SDDL, SDDL_REVISION_1, &sa.lpSecurityDescriptor, &sa.nLength)) ||
+			sshkey_to_blob(key, &pubkey_blob, &pubkey_blob_len) != 0 ||
+			((thumbprint = sshkey_fingerprint(key, SSH_FP_HASH_DEFAULT, SSH_FP_DEFAULT)) == NULL) ||
+			RegCreateKeyExW(user_root, SSH_KEYS_ROOT, 0, 0, 0, KEY_WRITE | KEY_WOW64_64KEY, &sa, &reg, NULL) != 0 ||
+			RegCreateKeyExA(reg, thumbprint, 0, 0, 0, KEY_WRITE | KEY_WOW64_64KEY, &sa, &sub, NULL) != 0 ||
+			RegSetValueExW(sub, NULL, 0, REG_BINARY, pubkey_blob, (DWORD)pubkey_blob_len) != 0 ||
+			RegSetValueExW(sub, L"pub", 0, REG_BINARY, pubkey_blob, (DWORD)pubkey_blob_len) != 0 ||
+			RegSetValueExW(sub, L"type", 0, REG_DWORD, (BYTE*)&key->type, 4) != 0 ||
+			RegSetValueExW(sub, L"comment", 0, REG_BINARY, canonical_provider, (DWORD)strlen(canonical_provider)) != 0) {
+			debug("failed to add key to store");
+			goto done;
+		}
+	}
```

Comment:
Aren't we missing `LocalFree(sa.lpSecurityDescriptor)`, `RegCloseKey(reg)`, and `RegCloseKey(sub)` in-between loops?